### PR TITLE
Improve isWithinView in common.js #5937

### LINKED ISSUE #5937: Improve isWithinView in common.js
The function `isWithinView()` in common.js does not work for all HTML elements; due to the behaviour of [`element.offsetTop`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/offsetTop), this will presently only work for elements that are a direct child of `#mainContent`. An example of how the method works for several elements is shown below (using console logs not present in the original method):

viewport

console

Ideally, this should be changed to support all HTML elements instead.

### PR DESCRIPTION
**Pull request by li-kai**
Fixes #5937

### CODE REVIEW

**Comment by li-kai** [comment]
Ok so this is weird. Tests passes in Chrome but not in Firefox.
<hr>
Ok traced it down to a rounding error in Firefox.

**Comment by li-kai** [comment]
@wkurniawan07 Can you review or assign someone to review? It's been a day. Thanks!

**Comment by junhaoyap** [comment]
@li-kai @wkurniawan07 I'll take a look when I wake up later

**Comment by junhaoyap** [comment]
2am isn't a good time if there are major changes to a function

File: src/main/webapp/js/common.js [hunk 1]
```diff
@@ -484,20 +484,22 @@ function getPointValue(s, ditchZero) {
...
+    var rect = baseElement.getBoundingClientRect();
+
+    var windowHeight = window.innerHeight || document.documentElement.clientHeight;
+    var windowWidth = window.innerWidth || document.documentElement.clientWidth;
     
-    return viewHeight >= elementHeight
-           ? viewTop <= elementTop && viewBottom >= elementBottom          // all within view
-           : viewTop <= elementTop && viewBottom >= elementTop             // top within view
-             || viewTop <= elementBottom && viewBottom >= elementBottom;   // btm within view
+    var tolerance = 0.25;
```
Comment (junhaoyap) [inline src/main/webapp/js/common.js]: Add a comment about what the tolerance means, even with the stack overflow, it takes a deep reading in order to understand and not all maintainers will easily understand it

Also, it seems that the way we are using tolerance here is slightly different, the answer that it is taken from uses it as a percentage while we use it as an absolute pixel, can you elaborate on why we chose 0.25 pixels?

File: src/test/javascript/CommonJsTest.js [hunk 2]
```diff
@@ -1,5 +1,29 @@
...
+        var testString = 'Element with style ' + style + (condition ? ' is within view' : ' is not within view');
+        assert.equal(isWithinView(testDiv), condition, testString);
+    }
+    assertWithStyle('top: 0%;', true);
+    assertWithStyle('top: 100%;', false);
+    assertWithStyle('bottom: 0%;', true);
+    assertWithStyle('bottom: 100%;', false);
+    assertWithStyle('top: 0%; left: 0%;', true);
+    assertWithStyle('top: 0%; left: -100%;', false);
+
```
Comment (junhaoyap) [inline src/test/javascript/CommonJsTest.js]: Don't see why there is a need for an empty line here

File: src/main/webapp/js/common.js (hunk 1 above)
Comment (li-kai) [inline src/main/webapp/js/common.js]: Sure, although it's arbitrary, I've taken a looked at the ranges of values that make the tests fail on firefox.
```
rect.left + rect.width
> 1620.000015258789
windowWidth
> 1620
```

So the difference is really minuscule, but the tests still fail. Having it as a percentage isn't very good as the range could be really big for 4K screens.

File: src/test/javascript/CommonJsTest.js (hunk 2 above)
Comment (li-kai) [inline src/test/javascript/CommonJsTest.js]: You're right, I'll remove that.

**Comment by li-kai** [comment]
Updated and implemented the fixes.

**Comment by junhaoyap** [comment]
cc @damithc @wkurniawan07

File: src/main/webapp/js/common.js [hunk 3]
```diff
@@ -484,20 +484,22 @@ function getPointValue(s, ditchZero) {
...
-           ? viewTop <= elementTop && viewBottom >= elementBottom          // all within view
-           : viewTop <= elementTop && viewBottom >= elementTop             // top within view
-             || viewTop <= elementBottom && viewBottom >= elementBottom;   // btm within view
+    var baseElement = $(element)[0]; // unwrap jquery element
+    var rect = baseElement.getBoundingClientRect();
+
+    var windowHeight = window.innerHeight || document.documentElement.clientHeight;
+    var windowWidth = window.innerWidth || document.documentElement.clientWidth;
+
+    var tolerance = 0.25; // range to prevent rounding errors
```
Comment (damithc) [inline src/main/webapp/js/common.js]: A more detailed explanation of how this `tolerance` works can be useful.

File: src/main/webapp/js/common.js [hunk 4]
```diff
@@ -484,20 +484,22 @@ function getPointValue(s, ditchZero) {
 /**
  * Checks if element is within browser's viewport.
  * @return true if it is within the viewport, false otherwise
+ * @see http://stackoverflow.com/q/123999
  */
 function isWithinView(element) {
```
Comment (damithc) [inline src/main/webapp/js/common.js]: Is this method used in production code? Sounds like a method used for testing.

File: src/test/javascript/CommonJsTest.js [hunk 5]
```diff
@@ -1,5 +1,28 @@
...
+        assert.equal(isWithinView(testDiv), condition, testString);
+    }
+    assertWithStyle('top: 0%;', true);
+    assertWithStyle('top: 100%;', false);
+    assertWithStyle('bottom: 0%;', true);
+    assertWithStyle('bottom: 100%;', false);
+    assertWithStyle('top: 0%; left: 0%;', true);
+    assertWithStyle('top: 0%; left: -100%;', false);
+    assertWithStyle('top: 0%; right: -100%;', false);
+    assertWithStyle('top: 0%; right: 0%;', true);
```
Comment (damithc) [inline src/test/javascript/CommonJsTest.js]: This too can benefit from a comment explaining how the tests work.

File: src/main/webapp/js/common.js (hunk 4 above)
Comment (li-kai) [inline src/main/webapp/js/common.js]: It is indeed used for testing by `/src/test/java/teammates/test/pageobjects/AppPage.java`

Comment (damithc) [inline src/main/webapp/js/common.js]: My understanding was code used solely for testing should not be in files used for production.

Comment (li-kai) [inline src/main/webapp/js/common.js]: There is another usage of it in `scrollToElement` which is indeed used in multiple places in production.

Comment (damithc) [inline src/main/webapp/js/common.js]: > There is another usage of it in scrollToElement in common.js which is indeed used in multiple places in production.

It's fine in that case. Thanks for checking.

**Comment by wkurniawan07** [comment]
@li-kai sorry to crash the party, but the red X here is a legitimate one :p

**Comment by li-kai** [comment]
@damithc Updated to fulfil requests

File: src/main/webapp/js/common.js [hunk 6]
```diff
@@ -484,20 +484,23 @@ function getPointValue(s, ditchZero) {
...
-    var elementTop = element.offsetTop;
-    var elementBottom = elementTop + elementHeight;
+    var $viewport = $(window);
     
-    return viewHeight >= elementHeight
-           ? viewTop <= elementTop && viewBottom >= elementBottom          // all within view
-           : viewTop <= elementTop && viewBottom >= elementTop             // top within view
-             || viewTop <= elementBottom && viewBottom >= elementBottom;   // btm within view
+    // subpixel range to prevent rounding errors
+    // e.g. rect.top = -0.0013 will still be considered to be within viewport
```
Comment (damithc) [inline src/main/webapp/js/common.js]: Should be a `/** ... */` comments instead?
Not clear why you used `0.0013` in the comment. 
I guess the tolerance is used to consider a rectangle that is slightly _bigger_ than the specified one. Is that the case? If so, may be explain it like that?
The unexplained term `subpixel range` doesn't seem to be necessary to the explanation.

File: src/main/webapp/js/common.js [hunk 7]
```diff
@@ -484,20 +484,23 @@ function getPointValue(s, ditchZero) {
...
+    var tolerance = 0.25;
+    return (
+        rect.top >= 0 - tolerance                           // below the top of viewport
+        && rect.left >= 0 - tolerance                       // within the left of viewport
+        && rect.right <= $viewport.width() + tolerance      // within the right of viewport
+        && rect.bottom <= $viewport.height() + tolerance    // above the bottom of viewport
```
Comment (damithc) [inline src/main/webapp/js/common.js]: Aligning comments like this is not recommended as it waste space and increase the blast radius of future changes.

**Comment by li-kai** [comment]
> Some cosmetic changes

Are the current comments clear now? I haven't thought of the "bigger box" analogy, that was a good suggestion.

**Pull request merged**